From 4608020edf3dbe00ec6ca83483b6a9513a189a2a Mon Sep 17 00:00:00 2001
From: "Lenzeder, Joerg" <joerg.lenzeder@intel.com>
Date: Thu, 6 May 2021 17:14:57 +0200
Subject: [PATCH] PONRTSYS-8943: Add changes from LGM PON QoS driver back to
 PRX

Adaptations to have same source for kernel versions 4.9 and 4.19
---
 drivers/net/datapath/pon_qos/pon_qos_tc_drr.c      |  26 +--
 .../net/datapath/pon_qos/pon_qos_tc_ds_tc_remap.c  |   2 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c |  10 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.h |   4 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_flower.c   | 193 ++++++++++++++++++++-
 drivers/net/datapath/pon_qos/pon_qos_tc_flower.h   |   2 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_lct.c      |   1 +
 drivers/net/datapath/pon_qos/pon_qos_tc_main.c     |  76 +++++++-
 drivers/net/datapath/pon_qos/pon_qos_tc_main.h     |  20 +--
 drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c   |  22 ++-
 drivers/net/datapath/pon_qos/pon_qos_tc_mqprio.c   |  11 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_pce.c      |   4 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_police.c   |  23 ++-
 drivers/net/datapath/pon_qos/pon_qos_tc_police.h   |   4 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_prio.c     |  34 ++--
 drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c     |  26 ++-
 drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h     |   3 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.c      |   9 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.h      |   2 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_red.c      |  25 +--
 drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c      |  22 ++-
 drivers/net/datapath/pon_qos/pon_qos_tc_trap.c     |  13 +-
 .../net/datapath/pon_qos/pon_qos_tc_vlan_filter.c  |   6 +-
 .../net/datapath/pon_qos/pon_qos_tc_vlan_filter.h  |   4 +-
 .../net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c |  72 ++++++--
 .../net/datapath/pon_qos/pon_qos_tc_vlan_prepare.h |   6 +-
 .../net/datapath/pon_qos/pon_qos_tc_vlan_storage.c |   1 +
 27 files changed, 501 insertions(+), 120 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_drr.c b/drivers/net/datapath/pon_qos/pon_qos_tc_drr.c
index 6dd8335870fb..74560fc4cea4 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_drr.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_drr.c
@@ -8,6 +8,7 @@
 #include <net/pkt_cls.h>
 #include <net/datapath_api.h>
 #include <net/datapath_api_qos.h>
+#include <linux/version.h>
 #include "pon_qos_tc_qos.h"
 #include "pon_qos_trace.h"
 
@@ -112,7 +113,6 @@ static int pon_qos_drr_queue_update(struct net_device *dev,
 }
 
 static int pon_qos_tc_drr_replace(struct net_device *dev,
-				  u32 handle,
 				  struct tc_drr_qopt_offload *opt,
 				  int port_id,
 				  int deq_idx)
@@ -132,7 +132,7 @@ static int pon_qos_tc_drr_replace(struct net_device *dev,
 		newp = true;
 	}
 
-	if (!TC_H_MIN(handle)) {
+	if (!TC_H_MIN(opt->handle)) {
 		netdev_dbg(dev, "%s: Config drr sched\n", __func__);
 		port->root_qdisc.port = port_id;
 		port->root_qdisc.deq_idx = deq_idx;
@@ -152,9 +152,9 @@ static int pon_qos_tc_drr_replace(struct net_device *dev,
 	/* This is a drr class so queue has to be configured on
 	 * an existing port.
 	 */
-	if (TC_H_MIN(handle) && !newp) {
+	if (TC_H_MIN(opt->handle) && !newp) {
 		netdev_dbg(dev, "%s: Config queue\n", __func__);
-		qdisc = pon_qos_qdisc_find(port, handle);
+		qdisc = pon_qos_qdisc_find(port, opt->handle);
 		if (!qdisc) {
 			netdev_err(dev, "%s: radix tree broken\n", __func__);
 			return -ENODEV;
@@ -246,7 +246,6 @@ static int pon_qos_tc_drr_qdisc_del(struct pon_qos_port *port,
 }
 
 static int pon_qos_tc_drr_destroy(struct net_device *dev,
-				  u32 handle,
 				  struct tc_drr_qopt_offload *opt)
 {
 	struct pon_qos_port *port = NULL;
@@ -265,7 +264,7 @@ static int pon_qos_tc_drr_destroy(struct net_device *dev,
 		return 0;
 	}
 
-	ret = pon_qos_tc_drr_queue_del(sch, handle);
+	ret = pon_qos_tc_drr_queue_del(sch, opt->handle);
 	if (ret < 0)
 		return ret;
 
@@ -286,13 +285,18 @@ static int pon_qos_tc_drr_destroy(struct net_device *dev,
 }
 
 int pon_qos_tc_drr_offload(struct net_device *dev,
-			   u32 handle,
-			   struct tc_to_netdev *tc,
+			   void *type_data,
 			   int port_id,
 			   int deq_idx)
 {
 	int err = 0;
-	struct tc_drr_qopt_offload *opt = tc->sch_drr;
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
+	struct tc_drr_qopt_offload *opt =
+		((struct tc_to_netdev *)type_data)->sch_drr;
+#else
+	struct tc_drr_qopt_offload *opt =
+		(struct tc_drr_qopt_offload *)type_data;
+#endif
 
 	ASSERT_RTNL();
 	netdev_dbg(dev, "DRR: offload starting\n");
@@ -302,7 +306,7 @@ int pon_qos_tc_drr_offload(struct net_device *dev,
 	case TC_DRR_REPLACE:
 		netdev_dbg(dev, "pid: %#x class/handle: %#x quantum: %#x\n",
 			   opt->parent, opt->handle, opt->set_params.quantum);
-		err = pon_qos_tc_drr_replace(dev, handle, opt,
+		err = pon_qos_tc_drr_replace(dev, opt,
 					     port_id, deq_idx);
 		if (err < 0) {
 			netdev_err(dev, "tc-drr replace failed\n");
@@ -313,7 +317,7 @@ int pon_qos_tc_drr_offload(struct net_device *dev,
 		netdev_dbg(dev,
 			   "destroy pid: %#x class/handle: %#x quantum: %#x\n",
 			   opt->parent, opt->handle, opt->set_params.quantum);
-		err = pon_qos_tc_drr_destroy(dev, handle, opt);
+		err = pon_qos_tc_drr_destroy(dev, opt);
 		if (err < 0) {
 			netdev_err(dev, "tc-drr destroy failed\n");
 			return err;
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_ds_tc_remap.c b/drivers/net/datapath/pon_qos/pon_qos_tc_ds_tc_remap.c
index 1c411ee7cdac..0c6665473a43 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_ds_tc_remap.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_ds_tc_remap.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
- * Copyright (c) 2020 MaxLinear, Inc.
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2019-2020 Intel Corporation
  *
  *****************************************************************************/
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
index 26eec69b36ff..7f61ddead492 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
@@ -1,12 +1,14 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 -2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
 #include <linux/list.h>
 #include <linux/list_sort.h>
 #include <linux/netdevice.h>
+#include <linux/version.h>
 #include <net/pkt_cls.h>
 #include <net/flow_dissector.h>
 #include <net/tc_act/tc_vlan.h>
@@ -355,12 +357,13 @@ static int dp_update(struct net_device *dev,
 }
 
 int pon_qos_tc_ext_vlan_add(struct net_device *dev,
-			    struct tc_cls_flower_offload *f)
+			    struct tc_cls_flower_offload *f,
+			    bool ingress)
 {
 	struct dp_tc_vlan dp_vlan = {0};
 	struct pon_qos_vlan_storage_node *node;
 	enum tc_flower_vlan_tag tag;
-	bool ingress = pon_qos_tc_flower_is_ingress(dev, f);
+
 	int err = 0;
 
 	netdev_dbg(dev, "Extended VLAN add start\n");
@@ -550,16 +553,19 @@ static void print_node(struct seq_file *file,
 
 	if (p->tag == TC_VLAN_UNTAGGED) {
 		struct dp_vlan0 *i;
+
 		list_for_each_entry(i, &p->rules, list)
 			print_rule(file, &i->outer, NULL, &i->act);
 	}
 	if (p->tag == TC_VLAN_SINGLE_TAGGED) {
 		struct dp_vlan1 *i;
+
 		list_for_each_entry(i, &p->rules, list)
 			print_rule(file, &i->outer, NULL, &i->act);
 	}
 	if (p->tag == TC_VLAN_DOUBLE_TAGGED) {
 		struct dp_vlan2 *i;
+
 		list_for_each_entry(i, &p->rules, list)
 			print_rule(file, &i->outer, &i->inner, &i->act);
 	}
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.h b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.h
index 56c8526b0cd3..95c9c2b1ff57 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.h
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: GPL-2.0 */
 /******************************************************************************
  *
+ * Copyright (c) 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -13,7 +14,8 @@ struct net_device;
 struct seq_file;
 
 int pon_qos_tc_ext_vlan_add(struct net_device *dev,
-			    struct tc_cls_flower_offload *f);
+			    struct tc_cls_flower_offload *f,
+			    bool ingress);
 
 int pon_qos_tc_ext_vlan_del(struct net_device *dev,
 			    void *vlan_storage,
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_flower.c b/drivers/net/datapath/pon_qos/pon_qos_tc_flower.c
index c4e95954550f..4fe9d8504c88 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_flower.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_flower.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
- * Copyright (c) 2021 MaxLinear, Inc.
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -11,6 +11,7 @@
 #include <net/tc_act/tc_police.h>
 #include <net/tc_act/tc_gact.h>
 #include <net/flow_dissector.h>
+#include <linux/version.h>
 #include "pon_qos_tc_flower.h"
 #include "pon_qos_tc_ext_vlan.h"
 #include "pon_qos_tc_vlan_filter.h"
@@ -135,6 +136,7 @@ static int pon_qos_tc_flower_storage_del(struct net_device *dev,
 	return 0;
 }
 
+#if (KERNEL_VERSION(4, 10, 0) > LINUX_VERSION_CODE)
 static bool is_type_mirred(struct net_device *dev,
 			   struct tc_cls_flower_offload *f)
 {
@@ -151,12 +153,50 @@ static bool is_type_mirred(struct net_device *dev,
 
 	return false;
 }
+#elif (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
+static bool is_type_mirred(struct net_device *dev,
+			   struct tc_cls_flower_offload *f)
+{
+	const struct tc_action *a;
+	LIST_HEAD(actions);
+
+	if (!tcf_exts_has_actions(f->exts))
+		return false;
+
+	tcf_exts_to_list(f->exts, &actions);
+	list_for_each_entry(a, &actions, list)
+		if (is_tcf_mirred_egress_redirect(a))
+			return true;
+
+	return false;
+}
+#else
+static bool is_type_mirred(struct net_device *dev,
+			   struct tc_cls_flower_offload *f)
+{
+	const struct tc_action *a;
+	int i;
+
+	if (!tcf_exts_has_actions(f->exts))
+		return false;
+
+	tcf_exts_for_each_action(i, a, f->exts)
+		if (is_tcf_mirred_egress_redirect(a))
+			return true;
+
+	return false;
+}
+#endif
 
 static bool is_type_ip_drop(struct net_device *dev,
 			    struct tc_cls_flower_offload *f)
 {
 	const struct tc_action *a;
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	LIST_HEAD(actions);
+#else
+	int i;
+#endif
 
 	if (!dissector_uses_key(f->dissector, FLOW_DISSECTOR_KEY_BASIC) ||
 	    !dissector_uses_key(f->dissector, FLOW_DISSECTOR_KEY_CONTROL))
@@ -166,11 +206,18 @@ static bool is_type_ip_drop(struct net_device *dev,
 	    !dissector_uses_key(f->dissector, FLOW_DISSECTOR_KEY_IPV6_ADDRS))
 		return false;
 
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	if (tc_no_actions(f->exts))
 		return false;
 
 	tcf_exts_to_list(f->exts, &actions);
 	list_for_each_entry(a, &actions, list)
+#else
+	if (!tcf_exts_has_actions(f->exts))
+		return false;
+
+	tcf_exts_for_each_action(i, a, f->exts)
+#endif
 		if (is_tcf_gact_shot(a))
 			return true;
 
@@ -181,7 +228,11 @@ static bool is_type_colmark(struct net_device *dev,
 			    struct tc_cls_flower_offload *f)
 {
 	/* This driver supports only single actions for now */
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 	if (tc_no_actions(f->exts))
+#else
+	if (!tcf_exts_has_actions(f->exts))
+#endif
 		return false;
 
 	return is_tcf_colmark(f->exts->actions[0]);
@@ -191,7 +242,11 @@ static bool is_type_police(struct net_device *dev,
 			    struct tc_cls_flower_offload *f)
 {
 	/* This driver supports only single actions for now */
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 	if (tc_no_actions(f->exts))
+#else
+	if (!tcf_exts_has_actions(f->exts))
+#endif
 		return false;
 
 	return is_tcf_police(f->exts->actions[0]);
@@ -202,6 +257,7 @@ static bool is_type_trap(struct net_device *dev,
 			 struct tc_cls_flower_offload *f)
 {
 	const struct tc_action *a;
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	LIST_HEAD(actions);
 
 	if (tc_no_actions(f->exts))
@@ -209,6 +265,14 @@ static bool is_type_trap(struct net_device *dev,
 
 	tcf_exts_to_list(f->exts, &actions);
 	list_for_each_entry(a, &actions, list)
+#else
+	int i;
+
+	if (!tcf_exts_has_actions(f->exts))
+		return false;
+
+	tcf_exts_for_each_action(i, a, f->exts)
+#endif
 		if (((struct tcf_gact *)a)->tcf_action == TC_ACT_TRAP)
 			return true;
 
@@ -285,6 +349,7 @@ flower_type_get(struct net_device *dev, struct tc_cls_flower_offload *f)
 
 static int pon_qos_tc_flower_replace(struct net_device *dev,
 				     u32 handle,
+				     bool ingress,
 				     struct tc_cls_flower_offload *f)
 {
 	enum pon_qos_tc_flower_type type;
@@ -294,18 +359,18 @@ static int pon_qos_tc_flower_replace(struct net_device *dev,
 
 	switch (type) {
 	case TC_TYPE_EXT_VLAN:
-		return pon_qos_tc_ext_vlan_add(dev, f);
+		return pon_qos_tc_ext_vlan_add(dev, f, ingress);
 	case TC_TYPE_VLAN_FILTER:
-		return pon_qos_tc_vlan_filter_add(dev, f);
+		return pon_qos_tc_vlan_filter_add(dev, f, ingress);
 	case TC_TYPE_QUEUE:
-		return pon_qos_tc_map(dev, f);
+		return pon_qos_tc_map(dev, f, ingress);
 	case TC_TYPE_IP_DROP:
 		return pon_qos_ip_drop_offload(dev, f, handle);
 	case TC_TYPE_MIRRED:
 		return pon_qos_mirred_offload(dev, f, handle);
 	case TC_TYPE_POLICE:
 	case TC_TYPE_COLMARK:
-		return pon_qos_police_offload(dev, f);
+		return pon_qos_police_offload(dev, f, ingress);
 	case TC_TYPE_TRAP:
 		return pon_qos_trap_offload(dev, f, handle);
 	default:
@@ -322,23 +387,41 @@ static int pon_qos_tc_flower_destroy(struct net_device *dev,
 	return pon_qos_tc_flower_storage_del(dev, handle, f);
 }
 
+#if (KERNEL_VERSION(4, 14, 0) < LINUX_VERSION_CODE)
+int pon_qos_tc_flower_offload(struct net_device *dev,
+			      u32 handle,
+			      bool ingress,
+			      void *type_data)
+#else
 int pon_qos_tc_flower_offload(struct net_device *dev,
 			      u32 handle,
-			      struct tc_cls_flower_offload *f)
+			      void *type_data)
+#endif
 {
 	int ret = 0;
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
+	struct tc_cls_flower_offload *f =
+		((struct tc_to_netdev *)type_data)->cls_flower;
+	bool ingress = TC_H_MIN(f->common.classid) != TC_H_MIN(TC_H_MIN_EGRESS);
+#else
+	struct tc_cls_flower_offload *f =
+		(struct tc_cls_flower_offload *)type_data;
+#endif
 
 	ASSERT_RTNL();
 	netdev_dbg(dev, "%s:start %d\n", __func__, ret);
+
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 	if (TC_H_MAJ(f->common.classid) != TC_H_MAJ(TC_H_CLSACT))
 		return -EOPNOTSUPP;
+#endif
 
 	switch (f->command) {
 	case TC_CLSFLOWER_REPLACE:
 	{
 		netdev_dbg(dev, "%s:replace:\n", __func__);
 		trace_pon_qos_tc_flower_enter(dev, f, flower_type_get(dev, f));
-		ret = pon_qos_tc_flower_replace(dev, handle, f);
+		ret = pon_qos_tc_flower_replace(dev, handle, ingress, f);
 		break;
 	}
 	case TC_CLSFLOWER_DESTROY:
@@ -362,3 +445,99 @@ int pon_qos_tc_flower_offload(struct net_device *dev,
 	trace_pon_qos_tc_flower_exit(dev, f, -1);
 	return ret;
 }
+
+#if (KERNEL_VERSION(4, 14, 0) < LINUX_VERSION_CODE)
+int pon_qos_tc_block_cb_ingress(enum tc_setup_type type,
+				void *type_data,
+				void *cb_priv)
+{
+	int ret = 0;
+	struct net_device *dev = (struct net_device *)cb_priv;
+
+	ASSERT_RTNL();
+	netdev_dbg(dev, "%s:start\n", __func__);
+
+	if (!tc_cls_can_offload_and_chain0(dev, type_data))
+		return -EOPNOTSUPP;
+
+	if (type == TC_SETUP_CLSFLOWER) {
+		struct tc_cls_flower_offload *cls_flower =
+			(struct tc_cls_flower_offload *)type_data;
+
+		ret = pon_qos_tc_flower_offload(dev,
+						cls_flower->handle,
+						true,
+						cls_flower);
+	}
+
+	netdev_dbg(dev, "%s:exit %d\n", __func__, ret);
+	return ret;
+}
+
+int pon_qos_tc_block_cb_egress(enum tc_setup_type type,
+			       void *type_data,
+			       void *cb_priv)
+{
+	int ret = 0;
+	struct net_device *dev = (struct net_device *)cb_priv;
+
+	ASSERT_RTNL();
+	netdev_dbg(dev, "%s:start\n", __func__);
+
+	if (!tc_cls_can_offload_and_chain0(dev, type_data))
+		return -EOPNOTSUPP;
+
+	if (type == TC_SETUP_CLSFLOWER) {
+		struct tc_cls_flower_offload *cls_flower =
+			(struct tc_cls_flower_offload *)type_data;
+
+		ret = pon_qos_tc_flower_offload(dev,
+						cls_flower->handle,
+						false,
+						cls_flower);
+	}
+
+	netdev_dbg(dev, "%s:exit %d\n", __func__, ret);
+	return ret;
+}
+
+int pon_qos_tc_block_offload(struct net_device *dev,
+			     void *type_data)
+{
+	int ret = 0;
+	tc_setup_cb_t *cb;
+	struct tc_block_offload *f = (struct tc_block_offload *)type_data;
+
+	ASSERT_RTNL();
+	netdev_dbg(dev, " tc block offload starting - binder type %d\n",
+		   f->binder_type);
+
+	if (f->binder_type == TCF_BLOCK_BINDER_TYPE_CLSACT_INGRESS)
+		cb = pon_qos_tc_block_cb_ingress;
+	else if (f->binder_type == TCF_BLOCK_BINDER_TYPE_CLSACT_EGRESS)
+		cb = pon_qos_tc_block_cb_egress;
+	else
+		return -EOPNOTSUPP;
+
+	switch (f->command) {
+	case TC_BLOCK_BIND:
+		netdev_dbg(dev, "CLS BIND\n");
+		ret = tcf_block_cb_register(f->block, cb, (void *)dev,
+					    (void *)dev, f->extack);
+		if (ret)
+			return ret;
+
+		return 0;
+	case TC_BLOCK_UNBIND:
+		netdev_dbg(dev, "CLS UNBIND\n");
+		tcf_block_cb_unregister(f->block, cb, (void *)dev);
+
+		return 0;
+	default:
+		return -EOPNOTSUPP;
+	}
+
+	netdev_dbg(dev, "%s:exit %d\n", __func__, ret);
+	return ret;
+}
+#endif
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_flower.h b/drivers/net/datapath/pon_qos/pon_qos_tc_flower.h
index 31274670e7cf..16568d25fd5e 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_flower.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_flower.h
@@ -1,7 +1,7 @@
 /* SPDX-License-Identifier: GPL-2.0 */
 /******************************************************************************
  *
- * Copyright (c) 2021 MaxLinear, Inc.
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_lct.c b/drivers/net/datapath/pon_qos/pon_qos_tc_lct.c
index 941ff60af024..9a6d4d8f6b7a 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_lct.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_lct.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_main.c b/drivers/net/datapath/pon_qos/pon_qos_tc_main.c
index 4c668af38ac4..dfc41f27ee9a 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_main.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_main.c
@@ -5,6 +5,8 @@
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
+#include <linux/module.h>
+#include <linux/version.h>
 #include <net/pkt_cls.h>
 #include <net/pon_qos_tc.h>
 #include <linux/jiffies.h>
@@ -18,6 +20,7 @@
 
 #define DRV_NAME	"mod_pon_qos"
 
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 int pon_qos_setup_tc(struct net_device *dev,
 		     u32 handle,
 		     __be16 protocol,
@@ -36,25 +39,25 @@ int pon_qos_setup_tc(struct net_device *dev,
 	case TC_SETUP_CLSFLOWER:
 		ret = pon_qos_tc_flower_offload(dev,
 						handle,
-						tc->cls_flower);
+						tc);
 		break;
 	case TC_SETUP_MQPRIO:
 		ret = pon_qos_tc_mqprio_offload(dev, handle, tc,
 						port_id, deq_idx);
 		break;
 	case TC_SETUP_QDISC_PRIO:
-		ret = pon_qos_tc_prio_offload(dev, handle, tc,
+		ret = pon_qos_tc_prio_offload(dev, tc,
 					      port_id, deq_idx);
 		break;
 	case TC_SETUP_DRR:
-		ret = pon_qos_tc_drr_offload(dev, handle, tc,
+		ret = pon_qos_tc_drr_offload(dev, tc,
 					     port_id, deq_idx);
 		break;
 	case TC_SETUP_QDISC_RED:
-		ret = pon_qos_tc_red_offload(dev, handle, tc);
+		ret = pon_qos_tc_red_offload(dev, tc);
 		break;
 	case TC_SETUP_TBF:
-		ret = pon_qos_tc_tbf_offload(dev, handle, tc);
+		ret = pon_qos_tc_tbf_offload(dev, tc);
 		break;
 	default:
 		netdev_err(dev, "offload type:%d not supported\n", tc->type);
@@ -66,14 +69,57 @@ int pon_qos_setup_tc(struct net_device *dev,
 
 	return ret;
 }
+#else
+int pon_qos_setup_tc(struct net_device *dev,
+		     enum tc_setup_type type,
+		     void *type_data,
+		     int port_id,
+		     int deq_idx)
+{
+	int ret = 0;
+
+	ASSERT_RTNL();
+	netdev_dbg(dev, "%s: start type %d\n", __func__, type);
+	switch (type) {
+	case TC_SETUP_QDISC_MQPRIO:
+		return pon_qos_tc_mqprio_offload(dev,
+			((struct tc_mqprio_qopt_offload *)type_data)->handle,
+			type_data, port_id, deq_idx);
+	case TC_SETUP_QDISC_PRIO:
+		return pon_qos_tc_prio_offload(dev, type_data,
+			port_id, deq_idx);
+	case TC_SETUP_QDISC_DRR:
+		return pon_qos_tc_drr_offload(dev, type_data,
+			port_id, deq_idx);
+	case TC_SETUP_QDISC_RED:
+		return pon_qos_tc_red_offload(dev, type_data);
+	case TC_SETUP_QDISC_TBF:
+		return pon_qos_tc_tbf_offload(dev, type_data);
+	case TC_SETUP_BLOCK:
+		return pon_qos_tc_block_offload(dev, type_data);
+	case TC_SETUP_QDISC_MQ:
+		/* ToDo: Support to be added */
+		ret = -EOPNOTSUPP;
+		break;
+	default:
+		netdev_err(dev, "offload type:%d not supported\n", type);
+		ret = -EOPNOTSUPP;
+		break;
+	}
+
+	return ret;
+}
+#endif
 EXPORT_SYMBOL(pon_qos_setup_tc);
 
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 int pon_qos_setup_tc_gen(struct net_device *dev,
 			 u32 handle,
 			 __be16 protocol,
 			 struct tc_to_netdev *tc)
 {
 	int ret;
+
 	ret = pon_qos_setup_tc(dev, handle, protocol, tc, -1, -1);
 	if (!ret)
 		return ret;
@@ -86,6 +132,26 @@ int pon_qos_setup_tc_gen(struct net_device *dev,
 #endif
 	return ret;
 }
+#else
+int pon_qos_setup_tc_gen(struct net_device *dev,
+			 enum tc_setup_type type,
+			 void *type_data)
+{
+	int ret;
+
+	ret = pon_qos_setup_tc(dev, type, type_data, -1, -1);
+	if (!ret)
+		return ret;
+/* TODO: this has to be removed when all features are implemented in this
+ * driver
+ */
+#if IS_ENABLED(CONFIG_QOS_MGR)
+	if (qos_mgr_hook_setup_tc)
+		return qos_mgr_hook_setup_tc(dev, type, type_data);
+#endif
+	return ret;
+}
+#endif
 
 static int __init pon_qos_init(void)
 {
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_main.h b/drivers/net/datapath/pon_qos/pon_qos_tc_main.h
index f5234153a592..4dc70d40dd30 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_main.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_main.h
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: GPL-2.0 */
 /******************************************************************************
  *
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -8,32 +9,31 @@
 #define _PON_QOS_TC_MAIN_
 
 int pon_qos_tc_drr_offload(struct net_device *dev,
-			   u32 handle,
-			   struct tc_to_netdev *tc,
+			   void *type_data,
 			   int port_id,
 			   int deq_idx);
 
 int pon_qos_tc_prio_offload(struct net_device *dev,
-			    u32 handle,
-			    struct tc_to_netdev *tc,
+			    void *type_data,
 			    int port_id,
 			    int deq_idx);
 
 int pon_qos_tc_mqprio_offload(struct net_device *dev,
 			      u32 handle,
-			      struct tc_to_netdev *tc,
+			      void *type_data,
 			      int port_id,
 			      int deq_idx);
 
 int pon_qos_tc_flower_offload(struct net_device *dev,
 			      u32 handle,
-			      struct tc_cls_flower_offload *f);
+			      void *type_data);
 
 int pon_qos_tc_red_offload(struct net_device *dev,
-			   u32 handle,
-			   struct tc_to_netdev *tc);
+			   void *type_data);
 
 int pon_qos_tc_tbf_offload(struct net_device *dev,
-			   u32 handle,
-			   struct tc_to_netdev *tc);
+			   void *type_data);
+
+int pon_qos_tc_block_offload(struct net_device *dev,
+			     void *type_data);
 #endif
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c b/drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c
index 668233eb3dcb..86d0a94ea624 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_mirred.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -14,6 +15,7 @@
 #include <net/flow_dissector.h>
 #include <net/pkt_cls.h>
 #include <net/tc_act/tc_mirred.h>
+#include <linux/version.h>
 #include "pon_qos_tc_flower.h"
 #include <net/tc_act/tc_gact.h>
 #include "pon_qos_tc_pce.h"
@@ -72,21 +74,35 @@ static int pon_qos_parse_act_mirred(struct net_device *dev,
 				    bool *drop_act)
 {
 	const struct tc_action *a;
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	LIST_HEAD(actions);
 	int ifindex;
+#else
+	int i;
+#endif
 	int ret = -EINVAL;
 
 	*drop_act = false;
 
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 	if (tc_no_actions(exts))
+#else
+	if (!tcf_exts_has_actions(exts))
+#endif
 		return -EINVAL;
 
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	tcf_exts_to_list(exts, &actions);
 	list_for_each_entry(a, &actions, list) {
 		if (is_tcf_mirred_redirect(a) && ret != 0) {
 			ifindex = tcf_mirred_ifindex(a);
 			*mirr_dev = dev_get_by_index(dev_net(dev),
 						     ifindex);
+#else
+	tcf_exts_for_each_action(i, a, exts) {
+		if (is_tcf_mirred_egress_redirect(a) && ret != 0) {
+			*mirr_dev = tcf_mirred_dev(a);
+#endif
 			if (!(*mirr_dev))
 				continue;
 			ret = 0;
@@ -202,7 +218,7 @@ static int pon_qos_tc_to_pce(struct net_device *dev,
 		return -ENODEV;
 	}
 
-	netdev_dbg(dev, "%s: using port: %d subif: %d\n",
+	netdev_dbg(dev, "%s: using port: %d subif_groupid: %d\n",
 		   __func__, dp_subif.port_id, dp_subif.subif_groupid);
 
 	pce_rule->pattern.bEnable = 1;
@@ -211,7 +227,7 @@ static int pon_qos_tc_to_pce(struct net_device *dev,
 	pce_rule->pattern.bSubIfIdEnable = 1;
 	pce_rule->pattern.nSubIfId = dp_subif.subif_groupid;
 
-	if (flt->proto == ETH_P_ALL) {
+	if (ntohs(flt->proto) == ETH_P_ALL) {
 		ret = pon_qos_all_to_pce(dev, flt, pce_rule);
 	} else if (eth_type_vlan(flt->proto)) {
 		ret = pon_qos_vlan_to_pce(dev, flt, pce_rule);
@@ -294,7 +310,9 @@ int pon_qos_mirred_offload(struct net_device *dev,
 	}
 
 err:
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	dev_put(mirr_dev);
+#endif
 
 	return ret;
 }
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_mqprio.c b/drivers/net/datapath/pon_qos/pon_qos_tc_mqprio.c
index da3de35f5fb5..20ae5a61bfe9 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_mqprio.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_mqprio.c
@@ -13,6 +13,7 @@
 #include <net/sch_generic.h>
 #include <net/datapath_api.h>
 #include <net/datapath_api_qos.h>
+#include <linux/version.h>
 #include "pon_qos_tc_qos.h"
 
 /* create qdisc tree */
@@ -161,12 +162,18 @@ static int pon_qos_tc_mqprio_destroy(struct net_device *dev,
 
 int pon_qos_tc_mqprio_offload(struct net_device *dev,
 			      u32 handle,
-			      struct tc_to_netdev *tc,
+			      void *type_data,
 			      int port_id,
 			      int deq_idx)
 {
 	int err = 0;
-	struct tc_mqprio_qopt *opt = tc->mqprio;
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
+	struct tc_mqprio_qopt *opt =
+		((struct tc_to_netdev *)type_data)->mqprio;
+#else
+	struct tc_mqprio_qopt *opt =
+		&(((struct tc_mqprio_qopt_offload *)type_data)->qopt);
+#endif
 
 	ASSERT_RTNL();
 	netdev_dbg(dev, "MQPRIO: offload starting\n");
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c b/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
index e8ad1d61b2e3..ef2524610779 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_pce.c
@@ -76,8 +76,8 @@ static int get_subif_data(struct net_device *dev,
 		if (ret != DP_SUCCESS)
 			return -ENODEV;
 
-		pr_debug("%s: portid: %u, subif: %u\n",
-			 __func__, dp_subif.port_id, dp_subif.subif_groupid);
+		pr_debug("%s: portid: %u, subif_groupid: %u\n", __func__,
+			 dp_subif.port_id, dp_subif.subif_groupid);
 		*portid = dp_subif.port_id;
 		*subif = dp_subif.subif_groupid;
 		break;
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_police.c b/drivers/net/datapath/pon_qos/pon_qos_tc_police.c
index 7673c00a6b45..c498a79352a6 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_police.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_police.c
@@ -1,12 +1,14 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
 
 #include <linux/netdevice.h>
 #include <linux/types.h>
+#include <linux/version.h>
 #include <net/pkt_cls.h>
 #include <net/tc_act/tc_colmark.h>
 #include <net/tc_act/tc_police.h>
@@ -72,22 +74,33 @@ static struct net_device *pon_qos_get_indev(struct net_device *dev,
 
 static bool pon_qos_police_parse_actions(struct tc_cls_flower_offload *f,
 					 struct policer_data *pd,
-					 dp_subif_t *subif)
+					 dp_subif_t *subif, bool ingress)
 {
 	struct dp_meter_cfg *meter_cfg = &pd->meter_cfg;
 	struct policer_data *p = NULL;
 	struct net_device *indev;
 	const struct tc_action *a;
+#if (KERNEL_VERSION(4, 15, 0) > LINUX_VERSION_CODE)
 	LIST_HEAD(actions);
-	bool ingress;
+#else
+	int i;
+#endif
 	u64 burst;
 
 	pd->acts = 0;
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 	if (tc_no_actions(f->exts))
+#else
+	if (!tcf_exts_has_actions(f->exts))
+#endif
 		return false;
 
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	tcf_exts_to_list(f->exts, &actions);
 	list_for_each_entry(a, &actions, list) {
+#else
+	tcf_exts_for_each_action(i, a, f->exts) {
+#endif
 		if (!(pd->acts & PON_QOS_COLMARK_ACTION) && is_tcf_colmark(a)) {
 			/* The enumerations in the TC colmark are matching
 			 * enumerations in the dp manager.
@@ -131,7 +144,6 @@ static bool pon_qos_police_parse_actions(struct tc_cls_flower_offload *f,
 	}
 	netdev_dbg(pd->dev, "%s: acts: %d\n", __func__, pd->acts);
 
-	ingress = pon_qos_tc_flower_is_ingress(pd->dev, f);
 	meter_cfg->dir = ingress ? DP_DIR_INGRESS : DP_DIR_EGRESS;
 
 	indev = pon_qos_get_indev(pd->dev, f);
@@ -197,7 +209,8 @@ static int pon_qos_police_config(struct policer_data *pd, bool en)
 }
 
 int pon_qos_police_offload(struct net_device *dev,
-			   struct tc_cls_flower_offload *f)
+			   struct tc_cls_flower_offload *f,
+			   bool ingress)
 {
 	struct policer_data *pd = NULL;
 	struct policer_data tmp = { 0 };
@@ -211,7 +224,7 @@ int pon_qos_police_offload(struct net_device *dev,
 	}
 
 	tmp.dev = dev;
-	if (!pon_qos_police_parse_actions(f, &tmp, &subif)) {
+	if (!pon_qos_police_parse_actions(f, &tmp, &subif, ingress)) {
 		netdev_err(dev, "policer or marker actions parse error.\n");
 		return -EOPNOTSUPP;
 	}
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_police.h b/drivers/net/datapath/pon_qos/pon_qos_tc_police.h
index 68bf416dc157..df0bb2ec9eb3 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_police.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_police.h
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -10,7 +11,8 @@
 #include <net/pkt_cls.h>
 
 int pon_qos_police_offload(struct net_device *dev,
-			   struct tc_cls_flower_offload *f);
+			   struct tc_cls_flower_offload *f,
+			   bool ingress);
 
 int pon_qos_police_unoffload(struct net_device *dev,
 			     void *meter_cfg_ptr, void *flags_ptr);
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_prio.c b/drivers/net/datapath/pon_qos/pon_qos_tc_prio.c
index a8a5c35d6e73..d246df586a18 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_prio.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_prio.c
@@ -8,6 +8,7 @@
 #include <net/pkt_cls.h>
 #include <net/datapath_api.h>
 #include <net/datapath_api_qos.h>
+#include <linux/version.h>
 #include "pon_qos_tc_qos.h"
 #include "pon_qos_trace.h"
 
@@ -75,7 +76,6 @@ static int pon_qos_tc_prio_update_tree(struct pon_qos_qdisc *sch,
 }
 
 static int pon_qos_tc_prio_replace(struct net_device *dev,
-				   u32 handle,
 				   struct tc_prio_qopt_offload *opt,
 				   int port_id,
 				   int deq_idx)
@@ -147,7 +147,6 @@ static int pon_qos_tc_prio_replace(struct net_device *dev,
 }
 
 static int pon_qos_tc_prio_destroy(struct net_device *dev,
-				   u32 handle,
 				   struct tc_prio_qopt_offload *opt)
 {
 	struct pon_qos_port *port = NULL;
@@ -172,13 +171,19 @@ static int pon_qos_tc_prio_destroy(struct net_device *dev,
 }
 
 int pon_qos_tc_prio_offload(struct net_device *dev,
-			    u32 handle,
-			    struct tc_to_netdev *tc,
+			    void *type_data,
 			    int port_id,
 			    int deq_idx)
 {
 	int err = 0;
-	struct tc_prio_qopt_offload *opt = tc->sch_prio;
+
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
+	struct tc_prio_qopt_offload *opt =
+		((struct tc_to_netdev *)type_data)->sch_prio;
+#else
+	struct tc_prio_qopt_offload *opt =
+		(struct tc_prio_qopt_offload *)type_data;
+#endif
 
 	ASSERT_RTNL();
 	netdev_dbg(dev, "PRIO: offload starting\n");
@@ -189,12 +194,12 @@ int pon_qos_tc_prio_offload(struct net_device *dev,
 	case TC_PRIO_REPLACE:
 		if (opt->replace_params.bands > PON_QOS_TC_MAX_Q) {
 			netdev_err(dev, "tc-prio - out of range bands %d\n",
-			   opt->replace_params.bands);
+				   opt->replace_params.bands);
 			return -EINVAL;
 		}
 		netdev_dbg(dev, "replace pid:%#x class/handle:%#x bands:%#x\n",
 			   opt->parent, opt->handle, opt->replace_params.bands);
-		err = pon_qos_tc_prio_replace(dev, handle, opt,
+		err = pon_qos_tc_prio_replace(dev, opt,
 					      port_id, deq_idx);
 		if (err < 0) {
 			netdev_err(dev, "tc-prio replace failed\n");
@@ -202,9 +207,9 @@ int pon_qos_tc_prio_offload(struct net_device *dev,
 		}
 		break;
 	case TC_PRIO_DESTROY:
-		netdev_dbg(dev, "destroy pid:%#x class/handle:%#x bands:%#x\n",
-			   opt->parent, opt->handle, opt->replace_params.bands);
-		err = pon_qos_tc_prio_destroy(dev, handle, opt);
+		netdev_dbg(dev, "destroy pid:%#x class/handle:%#x\n",
+			   opt->parent, opt->handle);
+		err = pon_qos_tc_prio_destroy(dev, opt);
 		if (err < 0) {
 			netdev_err(dev, "tc-prio destroy failed\n");
 			return err;
@@ -212,13 +217,18 @@ int pon_qos_tc_prio_offload(struct net_device *dev,
 		break;
 	case TC_PRIO_STATS:
 		/* TODO */
-		netdev_dbg(dev, "stats pid:%#x class/handle:%#x bands:%#x\n",
-			   opt->parent, opt->handle, opt->replace_params.bands);
+		netdev_dbg(dev, "stats pid:%#x class/handle:%#x\n",
+			   opt->parent, opt->handle);
 		return -EOPNOTSUPP;
 	case TC_PRIO_GRAFT:
 		/* TODO */
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 		netdev_dbg(dev, "graft pid:%#x class/handle:%#x bands:%#x\n",
 			   opt->parent, opt->handle, opt->replace_params.bands);
+#else
+		netdev_dbg(dev, "graft pid:%#x class/handle:%#x band:%#x\n",
+			   opt->parent, opt->handle, opt->graft_params.band);
+#endif
 		return -EOPNOTSUPP;
 	default:
 		break;
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
index 96b8b3ebb55a..1a0c2c1a114e 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
@@ -7,6 +7,7 @@
  *****************************************************************************/
 #include <net/pkt_cls.h>
 #include <linux/list.h>
+#include <linux/version.h>
 #include <net/datapath_api.h>
 #include <net/flow_dissector.h>
 #include <net/tc_act/tc_mirred.h>
@@ -36,16 +37,28 @@ static int pon_qos_tc_parse_flower_action(struct tcf_exts *exts,
 					  char *tc_cookie)
 {
 	const struct tc_action *a;
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	LIST_HEAD(actions);
+#else
+	int i;
+#endif
 	int act_nr = 0;
 	int act_ok_nr = 0;
 	int offset;
 
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 	if (tc_no_actions(exts))
+#else
+	if (!tcf_exts_has_actions(exts))
+#endif
 		return -EINVAL;
 
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	tcf_exts_to_list(exts, &actions);
 	list_for_each_entry(a, &actions, list) {
+#else
+	tcf_exts_for_each_action(i, a, exts) {
+#endif
 		if (a->ops && a->ops->type == TCA_ACT_GACT &&
 		    to_gact(a)->tcf_action == TC_ACT_OK)
 			act_ok_nr++;
@@ -72,7 +85,8 @@ static int pon_qos_tc_parse_flower_action(struct tcf_exts *exts,
 }
 
 static int pon_qos_parse_tc_flower(struct tc_cls_flower_offload *f,
-				   struct flower_cls_map **flt)
+				   struct flower_cls_map **flt,
+				   bool ingress)
 {
 	int ret = 0;
 
@@ -100,10 +114,7 @@ static int pon_qos_parse_tc_flower(struct tc_cls_flower_offload *f,
 	(*flt)->pref = f->common.prio >> 16;
 	(*flt)->proto = f->common.protocol;
 	(*flt)->classid = f->classid;
-	if (TC_H_MIN(f->common.classid) == TC_H_MIN(TC_H_MIN_EGRESS))
-		(*flt)->ingress = false;
-	else
-		(*flt)->ingress = true;
+	(*flt)->ingress = ingress;
 
 	/* Classification/Matching arguments parsing */
 	if (dissector_uses_key(f->dissector, FLOW_DISSECTOR_KEY_VLAN)) {
@@ -183,7 +194,8 @@ static int __pon_qos_tc_map(struct net_device *dev,
 	return 0;
 }
 
-int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f)
+int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f,
+		   bool ingress)
 {
 	struct flower_cls_map *map = NULL;
 	struct net_device *indev = NULL;
@@ -201,7 +213,7 @@ int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f)
 	if (ret < 0)
 		goto err_put_indev;
 
-	ret = pon_qos_parse_tc_flower(f, &map);
+	ret = pon_qos_parse_tc_flower(f, &map, ingress);
 	if (ret < 0)
 		goto err_put_indev;
 
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h
index 716089afee82..a71fa6740ada 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h
@@ -8,7 +8,8 @@
 #ifndef _PON_QOS_TC_QMAP_
 #define _PON_QOS_TC_QMAP_
 
-int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f);
+int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f,
+		   bool ingress);
 int pon_qos_tc_unmap(struct net_device *dev, void *list_node);
 int pon_qos_tc_classid_unmap(u32 classid);
 
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
index 1707e9f0a33f..e71f85679a63 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
@@ -6,6 +6,7 @@
  *
  *****************************************************************************/
 #include <linux/netdevice.h>
+#include <linux/version.h>
 #include <net/datapath_api.h>
 #include <net/datapath_api_qos.h>
 #include <net/pon_qos_tc.h>
@@ -1200,14 +1201,14 @@ int pon_qos_qdata_remove(struct net_device *dev, struct pon_qos_q_data *qid,
 }
 
 static void pon_qos_shaper_config_init(struct pon_qos_qdisc *sch,
-			       struct pon_qos_q_data *qdata,
-			       enum dp_shaper_cmd cmd,
-			       struct dp_shaper_conf *cfg)
+				       struct pon_qos_q_data *qdata,
+				       enum dp_shaper_cmd cmd,
+				       struct dp_shaper_conf *cfg)
 {
 	cfg->inst = sch->inst;
 	cfg->type = DP_NODE_QUEUE;
 	cfg->cmd = cmd;
-	cfg->id.q_id = (int) qdata->qid;
+	cfg->id.q_id = (int)qdata->qid;
 }
 
 int pon_qos_shaper_add(struct pon_qos_qdisc *sch,
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.h b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.h
index c84d299bc14d..cb04083a550b 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.h
@@ -1,7 +1,7 @@
 /* SPDX-License-Identifier: GPL-2.0 */
 /******************************************************************************
  *
- * Copyright (c) 2020 MaxLinear, Inc.
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_red.c b/drivers/net/datapath/pon_qos/pon_qos_tc_red.c
index 346559a07f6b..5c573063c8c5 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_red.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_red.c
@@ -9,6 +9,7 @@
 #include <net/flow_dissector.h>
 #include <net/datapath_api.h>
 #include <net/datapath_api_qos.h>
+#include <linux/version.h>
 #include "pon_qos_tc_qos.h"
 #include "pon_qos_trace.h"
 
@@ -139,7 +140,6 @@ static int wred_green_destroy(struct net_device *dev, u32 handle, u32 parent)
 }
 
 static int pon_qos_tc_red_replace(struct net_device *dev,
-				  u32 handle,
 				  struct tc_red_qopt_offload *opt)
 {
 	struct pon_qos_port *port = NULL;
@@ -184,7 +184,8 @@ static int pon_qos_tc_red_replace(struct net_device *dev,
 		netdev_dbg(dev, "tc-red qdisc configuring yellow\n");
 	}
 
-	ret = pon_qos_qdata_add(dev, qid, handle, opt->parent, type, destroy);
+	ret = pon_qos_qdata_add(dev, qid, opt->handle, opt->parent,
+				type, destroy);
 	if (ret < 0)
 		return ret;
 
@@ -192,7 +193,7 @@ static int pon_qos_tc_red_replace(struct net_device *dev,
 	if (ret < 0)
 		return ret;
 
-	ret = radix_tree_insert(&port->qdiscs, TC_H_MAJ(handle), qdisc);
+	ret = radix_tree_insert(&port->qdiscs, TC_H_MAJ(opt->handle), qdisc);
 	if (ret) {
 		netdev_err(dev, "%s: qdisc insertion to radix tree failed: %d\n",
 			   __func__, ret);
@@ -205,7 +206,6 @@ static int pon_qos_tc_red_replace(struct net_device *dev,
 }
 
 static int pon_qos_tc_red_destroy(struct net_device *dev,
-				  u32 handle,
 				  struct tc_red_qopt_offload *opt)
 {
 	struct pon_qos_port *port = NULL;
@@ -237,7 +237,7 @@ static int pon_qos_tc_red_destroy(struct net_device *dev,
 		}
 	}
 
-	ret = pon_qos_qdata_remove(dev, qid, handle, opt->parent);
+	ret = pon_qos_qdata_remove(dev, qid, opt->handle, opt->parent);
 
 	trace_pon_qos_tc_red_destroy(dev, opt, qdisc, qid);
 
@@ -245,12 +245,17 @@ static int pon_qos_tc_red_destroy(struct net_device *dev,
 }
 
 int pon_qos_tc_red_offload(struct net_device *dev,
-			   u32 handle,
-			   struct tc_to_netdev *tc)
+			   void *type_data)
 {
-	struct tc_red_qopt_offload *opt = tc->sch_red;
 	int err = 0;
 
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
+	struct tc_red_qopt_offload *opt =
+		((struct tc_to_netdev *)type_data)->sch_red;
+#else
+	struct tc_red_qopt_offload *opt =
+		(struct tc_red_qopt_offload *)type_data;
+#endif
 	ASSERT_RTNL();
 	netdev_dbg(dev, "RED: offload starting\n");
 
@@ -260,7 +265,7 @@ int pon_qos_tc_red_offload(struct net_device *dev,
 	case TC_RED_REPLACE:
 		netdev_dbg(dev, "replace pid:%#x class/handle:%#x\n",
 			   opt->parent, opt->handle);
-		err = pon_qos_tc_red_replace(dev, handle, opt);
+		err = pon_qos_tc_red_replace(dev, opt);
 		if (err < 0) {
 			netdev_err(dev, "tc-red replace failed\n");
 			return err;
@@ -269,7 +274,7 @@ int pon_qos_tc_red_offload(struct net_device *dev,
 	case TC_RED_DESTROY:
 		netdev_dbg(dev, "destroy pid:%#x class/handle:%#x\n",
 			   opt->parent, opt->handle);
-		err = pon_qos_tc_red_destroy(dev, handle, opt);
+		err = pon_qos_tc_red_destroy(dev, opt);
 		if (err < 0) {
 			netdev_err(dev, "tc-red destroy failed\n");
 			return err;
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c b/drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c
index cd86d9fb39d5..130d82322a70 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c
@@ -1,10 +1,12 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
 #include <net/pkt_cls.h>
+#include <linux/version.h>
 #include "pon_qos_tc_qos.h"
 #include "pon_qos_trace.h"
 
@@ -27,13 +29,12 @@ static int pon_tbf_remove(struct net_device *dev, u32 handle, u32 parent)
 		return -ENODEV;
 	}
 
-
 	qid = pon_qos_qdata_qid_get(dev, qdisc, parent);
 	if (!qid) {
 		ret = pon_qos_get_queue_by_handle(dev, parent, &qid);
 		if (ret < 0 || !qid) {
 			netdev_err(dev, "%s: handle not found err %d\n",
-				__func__, ret);
+				   __func__, ret);
 			return -EINVAL;
 		}
 	}
@@ -48,7 +49,7 @@ static int pon_tbf_remove(struct net_device *dev, u32 handle, u32 parent)
 }
 
 static int pon_qos_tc_tbf_replace(struct net_device *dev,
-			   struct tc_tbf_qopt_offload *opt)
+				  struct tc_tbf_qopt_offload *opt)
 {
 	struct pon_qos_qdisc *qdisc = NULL;
 	struct pon_qos_q_data *qid = NULL;
@@ -72,7 +73,7 @@ static int pon_qos_tc_tbf_replace(struct net_device *dev,
 		ret = pon_qos_get_queue_by_handle(dev, opt->parent, &qid);
 		if (ret < 0 || !qid) {
 			netdev_err(dev, "%s: handle not found err %d\n",
-				__func__, ret);
+				   __func__, ret);
 			return -EINVAL;
 		}
 	}
@@ -124,7 +125,7 @@ static int pon_qos_tc_tbf_destroy(struct net_device *dev,
 		ret = pon_qos_get_queue_by_handle(dev, opt->parent, &qid);
 		if (ret < 0 || !qid) {
 			netdev_err(dev, "%s: handle not found err %d\n",
-				__func__, ret);
+				   __func__, ret);
 			return -EINVAL;
 		}
 	}
@@ -133,11 +134,16 @@ static int pon_qos_tc_tbf_destroy(struct net_device *dev,
 }
 
 int pon_qos_tc_tbf_offload(struct net_device *dev,
-			   u32 handle,
-			   struct tc_to_netdev *tc)
+			   void *type_data)
 {
 	int err = 0;
-	struct tc_tbf_qopt_offload *opt = tc->sch_tbf;
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
+	struct tc_tbf_qopt_offload *opt =
+		((struct tc_to_netdev *)type_data)->sch_tbf;
+#else
+	struct tc_tbf_qopt_offload *opt =
+		(struct tc_tbf_qopt_offload *)type_data;
+#endif
 
 	trace_pon_qos_tc_tbf_enter(dev, opt);
 
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c b/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
index 548a84b28fb4..2120c2f34bca 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
@@ -1,13 +1,14 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
- * Copyright (c) 2020 MaxLinear, Inc.
+ * Copyright (c) 2020 - 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
 
 #include <linux/netdevice.h>
 #include <linux/types.h>
+#include <linux/version.h>
 #include <net/datapath_api.h>
 #include <net/pkt_cls.h>
 #include <net/datapath_api_qos.h>
@@ -161,7 +162,7 @@ static void pon_qos_trap_parse_key_basic(struct net_device *dev,
 		}
 		if (mask->n_proto) {
 			pce_rule->pattern.bEtherTypeEnable = 1;
-			pce_rule->pattern.nEtherType = key->n_proto;
+			pce_rule->pattern.nEtherType = ntohs(key->n_proto);
 			netdev_dbg(dev, "%s: bEtherTypeEnable = %d\n",
 				   __func__,
 				   pce_rule->pattern.bEtherTypeEnable);
@@ -169,8 +170,6 @@ static void pon_qos_trap_parse_key_basic(struct net_device *dev,
 				   __func__, pce_rule->pattern.nEtherType);
 		}
 	}
-
-	return;
 }
 
 static bool is_mac_set(GSW_PCE_rule_t *pce_rule)
@@ -284,8 +283,6 @@ static void pon_qos_trap_parse_key_eth(struct net_device *dev,
 			}
 		}
 	}
-
-	return;
 }
 
 static void pon_qos_trap_parse_key_icmp(struct net_device *dev,
@@ -487,11 +484,11 @@ int pon_qos_trap_unoffload(struct net_device *dev,
 	ret = dp_pce_rule_del(&pce_blk_info, pce_rule);
 	if (ret < 0) {
 		netdev_err(dev, "%s: dp_pce_rule_del failed err %d\n",
-			    __func__, ret);
+			   __func__, ret);
 		return ret;
 	}
 	netdev_dbg(dev, "%s: deleted pce rule with index %d\n", __func__,
-		    pce_rule->nIndex);
+		   pce_rule->nIndex);
 
 	list_del(&p->list);
 	kfree(p);
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.c b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.c
index c279dfaf1092..aa588f766de5 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.c
@@ -8,6 +8,7 @@
 #include <linux/list.h>
 #include <linux/list_sort.h>
 #include <linux/netdevice.h>
+#include <linux/version.h>
 #include <net/pkt_cls.h>
 #include <net/flow_dissector.h>
 #include <net/tc_act/tc_vlan.h>
@@ -178,12 +179,12 @@ static int dp_update(struct net_device *dev,
 }
 
 int pon_qos_tc_vlan_filter_add(struct net_device *dev,
-			       struct tc_cls_flower_offload *f)
+			       struct tc_cls_flower_offload *f,
+			       bool ingress)
 {
 	struct dp_tc_vlan dp_vlan = {0};
 	struct pon_qos_vlan_storage_node *node;
 	enum tc_flower_vlan_tag tag;
-	bool ingress = pon_qos_tc_flower_is_ingress(dev, f);
 	int err = 0;
 
 	netdev_dbg(dev, "VLAN filter add, flower: %p\n", f);
@@ -323,6 +324,7 @@ static struct net_device *pon_qos_tc_search_pmapper_on_list(int cvid,
 						      struct list_head *storage)
 {
 	struct pon_qos_vlan_storage_node *p;
+
 	list_for_each_entry(p, storage, list) {
 		rtnl_lock();
 		if (p->ingress == ingress &&
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.h b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.h
index 508a0475a353..cfd979351a98 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.h
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: GPL-2.0 */
 /******************************************************************************
  *
+ * Copyright (c) 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -9,7 +10,8 @@
 #define _PON_QOS_TC_VLAN_FILTER_
 
 int pon_qos_tc_vlan_filter_add(struct net_device *dev,
-			       struct tc_cls_flower_offload *f);
+			       struct tc_cls_flower_offload *f,
+			       bool ingress);
 
 int pon_qos_tc_vlan_filter_del(struct net_device *dev,
 			       void *vlan_storage,
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
index ac119a9ec045..823363eabfc8 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2020 -2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -10,6 +11,7 @@
 #include <net/tc_act/tc_gact.h>
 #include <net/tc_act/tc_mirred.h>
 #include <net/datapath_api_vlan.h>
+#include <linux/version.h>
 #include <uapi/linux/tc_act/tc_vlan.h>
 #include "pon_qos_tc_vlan_storage.h"
 #include "pon_qos_tc_vlan_prepare.h"
@@ -20,12 +22,6 @@
  * struct dp_vlan(0,1,2) structures.
  */
 
-bool pon_qos_tc_flower_is_ingress(struct net_device *dev,
-				  struct tc_cls_flower_offload *f)
-{
-	return TC_H_MIN(f->common.classid) != TC_H_MIN(TC_H_MIN_EGRESS);
-}
-
 void pon_qos_tc_dp_vlan_prepare(struct net_device *dev,
 				struct dp_tc_vlan *dp_vlan,
 				bool ingress)
@@ -59,7 +55,7 @@ pon_qos_tc_vlan_tag_get(struct net_device *dev,
 	if (eth_type_vlan(f->common.protocol))
 		return __pon_qos_tc_vlan_tag_get(dev, f);
 
-	if (f->common.protocol == ETH_P_ALL)
+	if (ntohs(f->common.protocol) == ETH_P_ALL)
 		return TC_VLAN_UNTAGGED;
 
 	if (dissector_uses_key(f->dissector, FLOW_DISSECTOR_KEY_BASIC)) {
@@ -93,8 +89,8 @@ static int pon_qos_tc_get_ethtype(struct net_device *dev,
 
 		if (mask->n_proto) {
 			netdev_dbg(dev, "%s:etype %#x\n",
-				   __func__, key->n_proto);
-			return key->n_proto;
+				   __func__, ntohs(key->n_proto));
+			return ntohs(key->n_proto);
 		}
 	}
 
@@ -122,10 +118,18 @@ static bool fwd_drop_action_parse(struct net_device *dev,
 				  struct dp_act_vlan *act)
 {
 	const struct tc_action *a;
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	LIST_HEAD(actions);
+#else
+	int i;
+#endif
 
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	tcf_exts_to_list(f->exts, &actions);
 	list_for_each_entry(a, &actions, list) {
+#else
+	tcf_exts_for_each_action(i, a, f->exts) {
+#endif
 		/* DROP and PASS make no sense together */
 		if (is_tcf_gact_shot(a)) {
 			act->act = DP_VLAN_ACT_DROP;
@@ -168,7 +172,7 @@ static int __vlan_parse(struct net_device *dev,
 			patt->vid = key->vlan_id;
 		if (mask->vlan_priority == 0x7)
 			patt->prio = key->vlan_priority;
-		patt->tpid = key->vlan_tpid;
+		patt->tpid = ntohs(key->vlan_tpid);
 	}
 
 	netdev_dbg(dev, "VLAN parsed, vid: %d, tpid: %x, prio: %d\n",
@@ -196,10 +200,18 @@ static void cookie_parse(struct net_device *dev,
 			 struct dp_act_vlan *act)
 {
 	const struct tc_action *a;
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	LIST_HEAD(actions);
+#else
+	int i;
+#endif
 
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	tcf_exts_to_list(f->exts, &actions);
 	list_for_each_entry(a, &actions, list) {
+#else
+	tcf_exts_for_each_action(i, a, f->exts) {
+#endif
 		if (a->act_cookie) {
 			u8 *p = a->act_cookie->data;
 			u64 hi = 0;
@@ -254,8 +266,13 @@ static void dscp_parse(struct net_device *dev,
 					f->key);
 
 	dscp_value = (key->tos & IP_TOS_TO_DSCP) >> 2;
+#if (KERNEL_VERSION(4, 15, 0) > LINUX_VERSION_CODE)
 	act->dscp_pcp_map[dscp_value] =
 		((struct tcf_vlan *)a)->tcfv_push_prio;
+#else
+	act->dscp_pcp_map[dscp_value] =
+		((struct tcf_vlan *)a)->vlan_p->tcfv_push_prio;
+#endif
 	act->dscp_pcp_map_valid[dscp_value] = true;
 	act->prio[act->push_n] = DERIVE_FROM_DSCP;
 	netdev_dbg(dev, "DSCP detected, value: %u, prio: %u\n",
@@ -267,7 +284,11 @@ static void vlan_action_modify_prio_parse(struct net_device *dev,
 					  enum tc_flower_vlan_tag tag,
 					  int *prio)
 {
+#if (KERNEL_VERSION(4, 15, 0) > LINUX_VERSION_CODE)
 	*prio = ((struct tcf_vlan *)a)->tcfv_push_prio;
+#else
+	*prio = ((struct tcf_vlan *)a)->vlan_p->tcfv_push_prio;
+#endif
 
 	if (*prio)
 		return;
@@ -284,8 +305,15 @@ static void vlan_action_vid_parse(struct net_device *dev,
 				  int *vid)
 {
 	struct tcf_vlan *v = to_vlan(a);
+#if (KERNEL_VERSION(4, 15, 0) <= LINUX_VERSION_CODE)
+	struct tcf_vlan_params *p;
+
+	p = rcu_dereference_bh(v->vlan_p);
 
+	if (p->tcfv_push_flags & ACTVLAN_PUSH_F_ID) {
+#else
 	if (v->tcfv_push_flags & ACTVLAN_PUSH_F_ID) {
+#endif
 		*vid = tcf_vlan_push_vid(a);
 		return;
 	}
@@ -302,17 +330,29 @@ static int vlan_action_parse(struct net_device *dev,
 			     enum tc_flower_vlan_tag tag)
 {
 	const struct tc_action *a;
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	LIST_HEAD(actions);
+#else
+	int i;
+#endif
 
+#if (KERNEL_VERSION(4, 14, 0) > LINUX_VERSION_CODE)
 	if (tc_no_actions(f->exts)) {
+#else
+	if (!tcf_exts_has_actions(f->exts)) {
+#endif
 		netdev_dbg(dev, "TC no actions!\n");
 		return -EIO;
 	}
 	netdev_dbg(dev, "TC actions count: %d\n",
 		   (f->exts)->nr_actions);
 
+#if (KERNEL_VERSION(4, 19, 0) > LINUX_VERSION_CODE)
 	tcf_exts_to_list(f->exts, &actions);
 	list_for_each_entry(a, &actions, list) {
+#else
+	tcf_exts_for_each_action(i, a, f->exts) {
+#endif
 		if (!is_tcf_vlan(a))
 			continue;
 
@@ -331,8 +371,14 @@ static int vlan_action_parse(struct net_device *dev,
 
 		if (tcf_vlan_action(a) == TCA_VLAN_ACT_PUSH) {
 			act->act |= DP_VLAN_ACT_PUSH;
+
+#if (KERNEL_VERSION(4, 15, 0) > LINUX_VERSION_CODE)
 			act->prio[act->push_n] =
 				((struct tcf_vlan *)a)->tcfv_push_prio;
+#else
+			act->prio[act->push_n] =
+				((struct tcf_vlan *)a)->vlan_p->tcfv_push_prio;
+#endif
 			netdev_dbg(dev, "PUSH action detected, push_n: %d, act: %x\n",
 				   act->push_n + 1, act->act);
 		} else if (tcf_vlan_action(a) == TCA_VLAN_ACT_MODIFY) {
@@ -345,7 +391,7 @@ static int vlan_action_parse(struct net_device *dev,
 		}
 
 		vlan_action_vid_parse(dev, a, tag, &act->vid[act->push_n]);
-		act->tpid[act->push_n] = tcf_vlan_push_proto(a);
+		act->tpid[act->push_n] = ntohs(tcf_vlan_push_proto(a));
 		act->dei[act->push_n] = 0;
 
 		netdev_dbg(dev, "PUSH action parsed, vid: %d, tpid: %x, prio: %d, dei: %d\n",
@@ -375,7 +421,7 @@ int pon_qos_tc_vlan_untagged_prepare(struct net_device *dev,
 				     struct dp_tc_vlan *dp_vlan,
 				     struct dp_vlan0 *rule)
 {
-	u32 proto = f->common.protocol;
+	u32 proto = ntohs(f->common.protocol);
 	bool drop;
 	int err;
 
@@ -410,7 +456,7 @@ int pon_qos_tc_vlan_untagged_flt_prepare(struct net_device *dev,
 					 struct dp_tc_vlan *dp_vlan,
 					 struct dp_vlan0 *rule)
 {
-	u32 proto = f->common.protocol;
+	u32 proto = ntohs(f->common.protocol);
 	int err;
 
 	vlan_pattern_default_set(&rule->outer);
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.h b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.h
index ff806a648fb3..e39a375322df 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_prepare.h
@@ -1,6 +1,7 @@
 /* SPDX-License-Identifier: GPL-2.0 */
 /******************************************************************************
  *
+ * Copyright (c) 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -8,6 +9,8 @@
 #ifndef _PON_QOS_TC_VLAN_PREPARE_
 #define _PON_QOS_TC_VLAN_PREPARE_
 
+struct tc_cls_flower_offload;
+
 enum tc_flower_vlan_tag {
 	TC_VLAN_UNKNOWN = 0,
 	TC_VLAN_UNTAGGED = 1,
@@ -15,9 +18,6 @@ enum tc_flower_vlan_tag {
 	TC_VLAN_DOUBLE_TAGGED = 3,
 };
 
-bool pon_qos_tc_flower_is_ingress(struct net_device *dev,
-				  struct tc_cls_flower_offload *f);
-
 void pon_qos_tc_dp_vlan_prepare(struct net_device *dev,
 				struct dp_tc_vlan *dp_vlan,
 				bool ingress);
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_storage.c b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_storage.c
index a0e98d11a1a9..c5dc62dfd96e 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_storage.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_storage.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
+ * Copyright (c) 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
